flink-tests/src/test/java/org/apache/flink/test/checkpointing/NotifyCheckpointAbortedITCase.java

```diff
@@ -0,0 +1,576 @@
+	private static final long SKIPPED_CHECKPOINT_ID = 1L;
+	private static final long ABORTED_CHECKPOINT_ID = 2L;
+	private static final long TEST_TIMEOUT = 60000;
+	private static final String StuckAsyncCheckpointMapName = "StuckAsyncCheckpointMap";
+	private static MiniClusterWithClientResource cluster;
+
+	private static Path checkpointPath;
+	private static String localRecoveryFolder;
+	private static StuckAsyncSnapshotStrategy stuckAsyncSnapshotStrategy;
+
+	@Parameterized.Parameter
+	public boolean unalignedCheckpointEnabled;
+
+	@Parameterized.Parameters(name = "unalignedCheckpointEnabled ={0}")
+	public static Collection<Boolean> parameter() {
+		return Arrays.asList(true, false);
+	}
+
+	@ClassRule
+	public static final TemporaryFolder TEMPORARY_FOLDER = new TemporaryFolder();
+
+	@Before
+	public void setup() throws Exception {
+		Configuration configuration = new Configuration();
+		configuration.setBoolean(CheckpointingOptions.LOCAL_RECOVERY, true);
+		localRecoveryFolder = TEMPORARY_FOLDER.newFolder().getAbsolutePath();
+		configuration.setString(CheckpointingOptions.LOCAL_RECOVERY_TASK_MANAGER_STATE_ROOT_DIRS, localRecoveryFolder);
+		configuration.setString(HighAvailabilityOptions.HA_MODE, TestingHAFactory.class.getName());
+
+		checkpointPath = new Path(TEMPORARY_FOLDER.newFolder().toURI());
+		cluster = new MiniClusterWithClientResource(
+			new MiniClusterResourceConfiguration.Builder()
+				.setConfiguration(configuration)
+				.setNumberTaskManagers(1)
+				.setNumberSlotsPerTaskManager(1).build());
+		cluster.before();
+
+		stuckAsyncSnapshotStrategy = new StuckAsyncSnapshotStrategy();
+		TestingCompletedCheckpointStore.addCheckpointLatch.reset();
+		TestingCompletedCheckpointStore.abortCheckpointLatch.reset();
+		StuckAsyncCheckpointMap.checkpointAbortedLatch.reset();
+		BeforeExecuteCheckpointSink.notifiedCheckpointLatch.reset();
+		BeforeExecuteCheckpointSink.snapshotIds.clear();
+	}
+
+	@After
+	public void shutdown() {
+		if (cluster != null) {
+			cluster.after();
+			cluster = null;
+		}
+
+	}
+
+	/**
+	 * Verify operator at different phase of checkpoint could act as expected when notified of checkpoint abortion.
+	 *
+	 * <p>The job would run with at least two checkpoints. The 1st checkpoint would fail to add checkpoint to store,
+	 * and we verify all local states stored in 1st checkpoint would then discarded. The 2nd checkpoint would decline
+	 * by 'DeclineSink'. Then we verify the async runnable future of 'StuckAsyncCheckpointMap' was canceled as expected,
+	 * and 'BeforeExecuteCheckpointSink' did not execute the sync phase of checkpoint.
+	 *
+	 * <p>The job graph looks like:
+	 * NormalSource --> keyBy --> NormalMap --> StuckAsyncCheckpointMap --> DeclineSink
+	 *                                  |
+	 *                                  |--> BarrierDelayMap -> BeforeExecuteCheckpointSink
+	 */
+	@Test(timeout = TEST_TIMEOUT)
+	public void testNotifyCheckpointAborted() throws Exception {
```

Comment:
This test does a great job of testing the functionality :+1: 

However, considering its complexity and unit-test coverage, I think a valid trade-off would be to simply test that operator is notified that the checkpoint is aborted.